### PR TITLE
Make parallel vector derive from VectorSpaceVector

### PR DESCRIPTION
**Pull request by kronbichler**
This is the first step to move `parallel::distributed::Vector` to `VectorSpaceVector`. Relates to #2535 and #745.

I now moved all code of the parallel vector to a class `LinearAlgebra::parallel::Vector`. @Rombur is this the right scheme to name things? I noticed that the inner `parallel` namespace is a bit confusing and I had to qualify some things, like the member of `ReadWriteVector`. Should we rather make it `LinearAlgebra::distributed::Vector` or `LinearAlgebra::MPI::Vector`?

Regarding the methods, I removed a few of the more obscure `operator==`, `equ` and `sadd` methods that we had in the class, whereas I put some others that are necessary when using this vector throughout the library still in the vector.

Another thing is that the vector still has access functionality as it was before (to keep the class synchronized with the rest of the library) but at some point at least parts of it could go away.

Finally, I did not put the import function in `ReadWriteVector` yet but that is easy to add.

The directly related tests are passing, the full test suite will run later today.

### CODE REVIEW

**Comment by Rombur** [comment]
@kronbichler You forgot to commit ```la_parallel_vector.h``` and ```la_parallel_vectors.templates.h```

I prefer ```LinearAlgebra::distributed::Vector``` instead of ```LinearAlgebra::parallel::Vector``` It makes it easier to understand what the vector does exactly.

**Comment by kronbichler** [comment]
@Rombur thanks, I forgot those files. They are now appended.

So right now `LinearAlgebra::distributed::Vector` is the favorite name for it. We can wait a bit to see what others think.

File: include/deal.II/base/partitioner.h [hunk 1]
```diff
@@ -58,7 +59,7 @@ namespace Utilities
      *
      * @author Katharina Kormann, Martin Kronbichler, 2010, 2011
      */
-    class Partitioner
+    class Partitioner : public ::dealii::LinearAlgebra::CommunicationPatternBase
```
Comment (Rombur) [inline include/deal.II/base/partitioner.h]: ```CommunicationPatternBase``` was made very small not by choice but because I wasn't sure what would be common between all the vectors. If you think some of the functions in ```Partitioner``` could be useful to the others vectors just move it to the base class.

File: include/deal.II/lac/la_parallel_vector.templates.h [hunk 2]
```diff
@@ -635,6 +720,47 @@ namespace parallel
 
     template <typename Number>
     void
+    Vector<Number>::import(const ReadWriteVector<Number>                  &V,
+                           VectorOperation::values                         operation,
+                           std_cxx11::shared_ptr<const CommunicationPatternBase> communication_pattern)
+    {
+      // If no communication pattern is given, create one. Otherwise, use the
+      // given one.
+      std_cxx11::shared_ptr<const Utilities::MPI::Partitioner> comm_pattern;
```
Comment (Rombur) [inline include/deal.II/lac/la_parallel_vector.templates.h]: There could also be a case where the communication pattern is not given but was already created and can be reused. I don't remember if it is what I did or not.

File: include/deal.II/lac/la_parallel_vector.templates.h [hunk 3]
```diff
@@ -635,6 +720,47 @@ namespace parallel
...
+            std_cxx11::dynamic_pointer_cast<const Utilities::MPI::Partitioner> (communication_pattern);
+          AssertThrow(comm_pattern != NULL,
+                      ExcMessage(std::string("The communication pattern is not of type ") +
+                                 "Utilities::MPI::Partitioner."));
+        }
+      LinearAlgebra::parallel::Vector<Number> tmp_vector(comm_pattern);
+
+      // fill entries from ReadWriteVector into the distributed vector,
+      // including ghost entries. this is not really efficient right now
+      // because indices are translated twice, once for
```
Comment (Rombur) [inline include/deal.II/lac/la_parallel_vector.templates.h]: Missing end of the comment?

File: include/deal.II/lac/la_parallel_vector.templates.h (hunk 2 above)
Comment (kronbichler) [inline include/deal.II/lac/la_parallel_vector.templates.h]: In the epetra vector you store a partitioner. We should introduce that optimization here as well but for the moment my aim was to discuss the interface, keep all backward functionality (that we should restrict somehow before the release) and see if this is a sustainable path. (The optimizations also include the remainder discussed in #2535 regarding fast change between this vector and a readwrite vector.)

File: include/deal.II/lac/la_parallel_vector.templates.h (hunk 3 above)
Comment (kronbichler) [inline include/deal.II/lac/la_parallel_vector.templates.h]: I will append that with my next patch when I move the vector to the `ReadWriteVector::distributed` namespace (if that is what comes out). For completeness already now: The two index translations are for nth_index_in_set and once for tmp_vector::operator().

File: include/deal.II/lac/la_parallel_vector.templates.h (hunk 2 above)
Comment (Rombur) [inline include/deal.II/lac/la_parallel_vector.templates.h]: OK. For me, the interface is good :-) I think that deprecating as many old functions as possible (like you did) is the way to go.

Comment (kronbichler) [inline include/deal.II/lac/la_parallel_vector.templates.h]: The only caveat is that with this change we need to release `ReadWriteVector` and `VectorSpaceVector` with the next release. But makes sense anyway.

There is two more thing (I will not manage to do that today), which is to teach ReadWriteVector to get data from a distributed vector and to add functions to copy from and to parallel PETSc/Trilinos vectors. These will be added but marked deprecated because we really want to promote using the ReadWriteVector. My runs revealed that I need to fix the following tests (this is a reminder to myself):
```
	790 - petsc/copy_to_dealvec.mpirun=4.debug (Failed)
	791 - petsc/copy_to_dealvec.mpirun=4.release (Failed)
	1046 - petsc/copy_to_dealvec_block.mpirun=4.debug (Failed)
	1047 - petsc/copy_to_dealvec_block.mpirun=4.release (Failed)
	6626 - mpi/parallel_block_vector_01.mpirun=1.debug (Failed)
	6627 - mpi/parallel_block_vector_01.mpirun=1.release (Failed)
	6668 - mpi/parallel_block_vector_01.mpirun=4.debug (Failed)
	6669 - mpi/parallel_block_vector_01.mpirun=4.release (Failed)
	7434 - mpi/parallel_block_vector_01.mpirun=10.debug (Failed)
	7435 - mpi/parallel_block_vector_01.mpirun=10.release (Failed)
	7632 - multigrid/step-16-04.debug (Failed)
	7633 - multigrid/step-16-04.release (Failed)
	7799 - trilinos/copy_to_dealvec.mpirun=4.debug (Failed)
	7800 - trilinos/copy_to_dealvec.mpirun=4.release (Failed)
	8143 - trilinos/copy_to_dealvec_block.mpirun=4.debug (Failed)
	8144 - trilinos/copy_to_dealvec_block.mpirun=4.release (Failed)
```
The problem in the step-16 test is that I do not have an operation `LA::distributed::Vector::add(Vector)` but only the += operation.

Comment (bangerth) [inline include/deal.II/lac/la_parallel_vector.templates.h]: Just remove/deprecate the function and make people use `+=`.

**Comment by Rombur** [comment]
@kronbichler do you need help on this?

**Comment by kronbichler** [comment]
I've been traveling during the last weeks so I did not get into it yet, but I will look into it towards the end of this week when I'm back.

**Comment by Rombur** [comment]
There is no rush. I dont' want you to feel pressured to do it right now. I am just offering some help ;-)

**Comment by kronbichler** [comment]
Now this is ready for review again. I changed the namespace to `LinearAlgebra::distributed::Vector` and added the exchange to Trilinos and PETSc vectors and `ReadWriteVector`. I did not introduce performance optimizations for re-using the communication pattern yet. Maybe @Rombur you can help me with that in a follow-up patch (or we leave it until somebody needs it)?

**Comment by kronbichler** [comment]
There is one more item to discuss: We previously supported operations such as `is_non_negative`, `operator==` and `operator!=`. I do not think we should have them in the general vector interfaces and thus, I did not provide the functionality. This breaks backwards compatibility. I could still introduce it here and deprecate it but I don't really like that because I want the vector to also work with complex numbers and thus, I would need quite a bit of unnecessary code.

If we agree on that, I will also carry this change over to the block vector. @Rombur: I guess we want to call it `LinearAlgebra::distributed::BlockVector` or do you have other suggestions?

Finally, I will then also replace `parallel::distributed::Vector` in the test suite with `LinearAlgebra::distributed::BlockVector`.

**Comment by bangerth** [comment]
I think it makes sense to drop the 3 functions you mention.

File: include/deal.II/lac/la_parallel_vector.h [hunk 4]
```diff
@@ -0,0 +1,1434 @@
...
+     * <li> Besides the usual global access operator() it is also possible to
+     * access vector entries in the local index space with the function @p
+     * local_element(). Locally owned indices are placed first, [0,
+     * local_size()), and then all ghost indices follow after them
+     * contiguously, [local_size(), local_size()+n_ghost_entries()).
+     * </ul>
+     *
+     * Functions related to parallel functionality:
+     * <ul>
+     * <li> The function <code>compress()</code> goes through the data
```
Comment (Rombur) [inline include/deal.II/lac/la_parallel_vector.h]: Do we still use compress with the new interface? The other vectors don't need it anymore. Do you want to keep both interfaces or should this be deprecated?

File: source/lac/trilinos_epetra_communication_pattern.cc [hunk 5]
```diff
@@ -45,7 +45,7 @@ namespace LinearAlgebra
       comm = std_cxx11::make_shared<const MPI_Comm>(communicator);
 
       Epetra_Map vector_space_vector_map = vector_space_vector_index_set.make_trilinos_map(*comm,
-                                           false);
+                                           true);
```
Comment (Rombur) [inline source/lac/trilinos_epetra_communication_pattern.cc]: Why did you change this? VectorSpaceVector are not supposed to have overlapping dofs because this won't be allowed in Tpetra

**Comment by Rombur** [comment]
- I will add the optimization with the sparsity pattern once this has been merged.
 - ```LinearAlgebra::distributed::BlockVector``` sounds good.
 - Is this ready to be merged or still a work-in-progress?

File: include/deal.II/lac/la_parallel_vector.h (hunk 4 above)
Comment (kronbichler) [inline include/deal.II/lac/la_parallel_vector.h]: Right now the `compress()` and `update_ghost_values()` methods are still needed. On the one hand because they are as of now still doing the actual data exchange in the two import() methods of `ReadWriteVector` and the current vector.  I plan to address that in a later commit when I switch to lightweight conversion without copy of the locally owned range as discussed in #2535 (this includes also `operator()`, `local_element()`, and `begin()/end()`).

File: source/lac/trilinos_epetra_communication_pattern.cc (hunk 5 above)
Comment (kronbichler) [inline source/lac/trilinos_epetra_communication_pattern.cc]: I merely did this because of the test `tests/trilinos/copy_to_dealvec.cc` includes a case where the trilinos vector is ghosted. I agree that it was a bad idea to change this. I will rather disallow this in the test. (Breaking a niche functionality like this straight away without a transition period with deprecated functions seems acceptable, just as we do for the `operator==`, `operator!=`, and `is_non_negative` methods.)

**Comment by kronbichler** [comment]
Give me an hour or two to finalize this patch by moving the parallel block vector to the same namespace as well and replace all occurrences of `parallel::distributed::Vector` in the library and testsuite by `LinearAlgebra::distributed::Vector`. Finally, I will also write a changelog.

(The next step are then the layered index set according to #2535, which will hopefully allow me to remove the access functionality of this class yet retain the overall efficiency of MatrixFree loops.)

File: include/deal.II/lac/la_parallel_block_vector.h [hunk 6]
```diff
@@ -0,0 +1,599 @@
...
+
+      /**
+       * Divide the entire vector by a fixed factor.
+       */
+      virtual BlockVector<Number> &operator/= (const Number factor);
+
+      /**
+       * Add the vector @p V to the present one.
+       */
+      virtual BlockVector<Number> &operator+= (const VectorSpaceVector<Number> &V);
```
Comment (Rombur) [inline include/deal.II/lac/la_parallel_block_vector.h]: Does this really make sense? I am wondering if this class should really derive from ```VectorSpaceVector```.

File: include/deal.II/lac/la_parallel_block_vector.templates.h [hunk 7]
```diff
@@ -0,0 +1,770 @@
...
+    BlockVector<Number>::~BlockVector ()
+    {}
+
+
+
+    template <typename Number>
+    BlockVector<Number> &
+    BlockVector<Number>::operator = (const value_type s)
+    {
+
```
Comment (Rombur) [inline include/deal.II/lac/la_parallel_block_vector.templates.h]: Extra blank line

File: include/deal.II/lac/la_parallel_block_vector.templates.h [hunk 8]
```diff
@@ -0,0 +1,770 @@
...
+
+    template <typename Number>
+    void
+    BlockVector<Number>::compress (::dealii::VectorOperation::values operation)
+    {
+      // start all requests for all blocks before finishing the transfers as
+      // this saves repeated synchronizations
+      for (unsigned int block=0; block<this->n_blocks(); ++block)
+        this->block(block).compress_start(block*10 + 8273, operation);
```
Comment (Rombur) [inline include/deal.II/lac/la_parallel_block_vector.templates.h]: Can you add a comment for the 8273.

File: include/deal.II/lac/la_parallel_block_vector.templates.h [hunk 9]
```diff
@@ -0,0 +1,770 @@
...
+    }
+
+
+
+    template <typename Number>
+    void
+    BlockVector<Number>::update_ghost_values () const
+    {
+      for (unsigned int block=0; block<this->n_blocks(); ++block)
+        this->block(block).update_ghost_values_start(block*10 + 9923);
```
Comment (Rombur) [inline include/deal.II/lac/la_parallel_block_vector.templates.h]: Same thing.

File: include/deal.II/lac/la_parallel_block_vector.h (hunk 6 above)
Comment (Rombur) [inline include/deal.II/lac/la_parallel_block_vector.h]: Never mind. After seeing the implementation, I guess it makes.

**Comment by Rombur** [comment]
Looks good :+1:

File: include/deal.II/lac/la_parallel_block_vector.h (hunk 6 above)
Comment (kronbichler) [inline include/deal.II/lac/la_parallel_block_vector.h]: I agree it is a bit confusing to have both the global block vector as well as each individual block be `VectorSpaceVector` types. But it makes sense in my opinion for the case where I want to build block preconditioners for the purpose of solving. I want to have both a global inner product, vector addition, and so on, inside an iterative solver, i.e., I need to satisfy the `VectorSpaceVector` interface (even though I'm not sure if we ever will implement the import function; I currently throw an exception there). At the same time, access to blocks individually is needed when multiplying with block matrices or preconditioning where I might apply an inner solver, which in turn needs a vector from `VectorSpaceVector`.

**Comment by kronbichler** [comment]
The current version passes the test suite (except for the usual failures).

File: include/deal.II/lac/la_parallel_block_vector.h (hunk 6 above)
Comment (Rombur) [inline include/deal.II/lac/la_parallel_block_vector.h]: Yes, I agree. I will do the same for Trilinos Vector.

**Comment by shakirbsm** [comment]
With these changes I am not able to build dealii as it breaks. In my case I am not using mpi and p4est or trillinos.

[detailedlog.txt](https://github.com/dealii/dealii/files/316642/detailedlog.txt)
[error.txt](https://github.com/dealii/dealii/files/316645/error.txt)

**Comment by kronbichler** [comment]
@shakirbsm I'm on it. I just need to check that it doesn't break anything else.

**Comment by Rombur** [comment]
@shakirbsm this should be fixed now. Thanks for the report!!

**Comment by shakirbsm** [comment]
@Rombur it still breaks as iomanip header is missing in la_parallel_vector.h.

**Comment by shakirbsm** [comment]
@kronbichler @Rombur It even breaks after adding that header and the error is in file
[error1.txt](https://github.com/dealii/dealii/files/318484/error1.txt)

**Comment by kronbichler** [comment]
@shakirbsm Thanks for the info. It took me a while to reproduce on my systems because I somehow missed the violating configuration. I have a patch.

**Comment by kronbichler** [comment]
See #2689.
